cpp/src/arrow/flight/client_cookie_middleware.cc

```diff
@@ -0,0 +1,327 @@
+};
+}  // end of anonymous namespace
+
+namespace arrow {
+namespace flight {
+
+using CookieCache = std::map<std::string, Cookie, CaseInsensitiveComparator>;
+using CookieHeaderPair =
+    const std::pair<CallHeaders::const_iterator, CallHeaders::const_iterator>&;
+/// \brief Client-side middleware for sending/receiving HTTP cookies.
+class ClientCookieMiddlewareFactory : public ClientMiddlewareFactory {
+ public:
+  void StartCall(const CallInfo& info, std::unique_ptr<ClientMiddleware>* middleware) {
+    ARROW_UNUSED(info);
+    *middleware = std::unique_ptr<ClientMiddleware>(new ClientCookieMiddleware(*this));
+  }
+
+ private:
+  class ClientCookieMiddleware : public ClientMiddleware {
+   public:
+    explicit ClientCookieMiddleware(ClientCookieMiddlewareFactory& factory)
+        : factory_(factory) {}
+
+    void SendingHeaders(AddCallHeaders* outgoing_headers) override {
+      const std::string& cookie_string = factory_.GetValidCookiesAsString();
+      if (!cookie_string.empty()) {
+        outgoing_headers->AddHeader("cookie", cookie_string);
+      }
+    }
+
+    void ReceivedHeaders(const CallHeaders& incoming_headers) override {
+      CookieHeaderPair cookie_header_values = incoming_headers.equal_range("set-cookie");
+      factory_.UpdateCachedCookies(cookie_header_values);
+    }
+
+    void CallCompleted(const Status& status) override {}
+
+   private:
+    ClientCookieMiddlewareFactory& factory_;
+  };
+
+  /// \brief Retrieve the cached cookie values as a string.
+  ///
+  /// \return a string that can be used in a Cookie header representing the cookies that
+  /// have been cached.
+  std::string GetValidCookiesAsString() {
+    const std::lock_guard<std::mutex> guard(mutex_);
+
+    DiscardExpiredCookies();
+    if (cookie_cache_.empty()) {
+      return "";
+    }
+
+    std::string cookie_string = cookie_cache_.begin()->second.AsCookieString();
+    for (auto it = (++cookie_cache_.begin()); cookie_cache_.end() != it; ++it) {
+      cookie_string += "; " + it->second.AsCookieString();
+    }
+    return cookie_string;
+  }
+
+  /// \brief Updates the cache of cookies with new Set-Cookie header values.
+  ///
+  /// \param header_values The range representing header values.
+  void UpdateCachedCookies(CookieHeaderPair header_values) {
+    const std::lock_guard<std::mutex> guard(mutex_);
+
+    for (auto it = header_values.first; it != header_values.second; ++it) {
+      const util::string_view& value = it->second;
+      Cookie cookie = Cookie::parse(value);
+
+      // Cache cookies regardless of whether or not they are expired. The server may have
+      // explicitly sent a Set-Cookie to expire a cached cookie.
+      std::pair<CookieCache::iterator, bool> insertable =
+          cookie_cache_.insert({cookie.cookie_name_, cookie});
+
+      // Force overwrite on insert collision.
+      if (!insertable.second) {
+        insertable.first->second = cookie;
+      }
+    }
+
+    DiscardExpiredCookies();
```

Comment:
Is there any need to discard here if we have to discard before we send cookies anyways?